### ISSUE TITLE
componentDidMount is called any time `this.setState` is called

### ISSUE THREAD

**Issue by selbekk** [airbnb/enzyme#1766]
**Describe the bug**
`componentDidMount` is called on any call to `this.setState`. The bug was introduced in 3.4.2.

**To Reproduce**
Here's a sample component
```js
import React from 'react';
import { shallow } from 'enzyme';

class TestComponent extends React.Component {
    state = {}
    componentDidMount() {
        this.props.onMount();
    }
    render() {
        return <button onClick={() => this.setState({ someState: 'some new state' })} />
    }
}

describe('componentDidMount', () => {
    it('will call componentDidMount on every call to setState', () => {
        const spy = jest.fn();
        const wrapper = shallow(<TestComponent onMount={spy} />);

        expect(spy).toHaveBeenCalledTimes(1);
        wrapper.find('button').simulate('click'); // updating state
        expect(spy).toHaveBeenCalledTimes(1);
    });
})
```

**Expected behavior**
I'd expect `componentDidMount` only to be called when I mount the component for the first time, or re-mount it explicitly. Instead, it calls it every time `this.setState` is called.

**Desktop (please complete the following information):**
 - Linux
 - Chrome / not applicable
 - 3.4.2 and 3.4.3 and latest?

Relates to https://github.com/airbnb/enzyme/issues/1756

**Comment by Android3000** [comment]
Is this possibly just 3.4.3? We just updated from 3.4.2 and noticed some previously passing tests that were failing due to componentDidMount being unexpectedly called, even in a shallow test with disableLifecycleMethods: true.

**Comment by ljharb** [comment]
cc @koba04

**Comment by monken** [comment]
This will be fixed by #1768, see #1765
